src/main/scala/com/fulcrumgenomics/basecalling/ExtractBasecallingParams.scala

```diff
@@ -0,0 +1,191 @@
+
+  override def execute(): Unit = {
+    val params = BasecallingParams.from(sampleSheet=SampleSheet(sampleSheet=sampleSheet), lane=lane, output=output)
+    logger.info(s"Barcode params written to: ${params.barcodeFile}")
+    logger.info(s"Library params written to: ${params.libraryParamsFile}")
+  }
+}
+
+
+/** A small class to store where the basecalling parameter and BAM files are located. */
+case class BasecallingParams(barcodeFile: FilePath, libraryParamsFile: FilePath, bams: Seq[PathToBam])
+
+/** Contains methods useful for extractin basecalling paramter information. */
+object BasecallingParams {
+
+  /** Extracts sample and library information from an sample sheet for a given lane.  The resulting files are used
+    * to run Picard's ExtractIlluminaBarcodes, CollectIlluminaBasecallingMetrics, and IlluminaBasecallsToSam. See
+    * [[ExtractBasecallingParams]] for more information.
+    *
+    * @param sampleSheet the Illumina Experiment Manager sample sheet.
+    * @param lane the lane number.
+    * @param output the output directory where the barcode, library param, and BAM files should be written.
+    */
+  def from(sampleSheet: SampleSheet, lane: Int, output: DirPath): BasecallingParams = {
+    require(0 < lane)
+
+    val barcodeLines      = new ListBuffer[String]()
+    val libraryParamLines = new ListBuffer[String]()
+    val samples           = sampleSheet.filter(_.lane.forall(_ == lane)).toSeq
+    val barcodeFile       = barcodeFileFrom(output=output, lane=lane)
+    val libraryParamsFile = libraryParamsFileFrom(output=output, lane=lane)
+
+    // Dual vs single indexed
+    val dualIndexed = samples.exists(_.i5IndexBases.nonEmpty)
+
+    // Check for project/description on the sample
+    val includeDescription    = samples.exists(s => s.project.nonEmpty || s.description.nonEmpty)
+    val standardLibraryParams = Seq("OUTPUT", "SAMPLE_ALIAS", "LIBRARY_NAME") ++ (if (includeDescription) Seq("DS") else Seq.empty)
+
+    // Headers
+    if (dualIndexed) {
+      barcodeLines += Seq("barcode_sequence_1", "barcode_sequence_2", "barcode_name", "library_name").mkString("\t")
+      libraryParamLines += (Seq("BARCODE_1", "BARCODE_2") ++ standardLibraryParams).mkString("\t")
+    }
+    else {
+      barcodeLines += Seq("barcode_sequence_1", "barcode_name", "library_name").mkString("\t")
+      libraryParamLines += (Seq("BARCODE_1") ++ standardLibraryParams).mkString("\t")
+    }
+
+    // Samples
+    val bams = samples.map { sample =>
+      // Get the library id
+      val libraryId = sample.libraryId.getOrElse(unreachable(s"No library found for sample: ${sample.sampleName}"))
+
+      // Get the barcodes
+      val barcodes = (sample.i7IndexBases, sample.i5IndexBases) match {
+        case (Some(barcode1Bases), Some(barcode2Bases)) => Seq(barcode1Bases, barcode2Bases)
+        case (Some(barcode1Bases), None) => Seq(barcode1Bases)
+        case _ => unreachable(s"No i7 index bases found for sample: ${sample.sampleName}")
```

Comment:
Is the genuinely unreachable?  Or is it possible given a poorly setup sample sheet?